src/fsharp/lexhelp.fs

```diff
@@ -305,48 +305,44 @@ module Keywords =
         keywordList |> List.map (fun (_, w, _) -> w) 
 
     let keywordTable = 
-        // TODO: this doesn't need to be a multi-map, a dictionary will do
         let tab = System.Collections.Generic.Dictionary<string,token>(100)
-        for (_mode,keyword,token) in keywordList do tab.Add(keyword,token)
+        for _,keyword,token in keywordList do 
+            tab.Add(keyword,token)
         tab
         
     let KeywordToken s = keywordTable.[s]
 
-    /// ++GLOBAL MUTABLE STATE. Note this is a deprecated, undocumented command line option anyway, we can ignore it.
-    let mutable permitFsharpKeywords = true
-
     let IdentifierToken args (lexbuf:UnicodeLexing.Lexbuf) (s:string) =
         if IsCompilerGeneratedName s then 
             warning(Error(FSComp.SR.lexhlpIdentifiersContainingAtSymbolReserved(), lexbuf.LexemeRange));
         args.resourceManager.InternIdentifierToken s
 
     let KeywordOrIdentifierToken args (lexbuf:UnicodeLexing.Lexbuf) s =
-        if not permitFsharpKeywords && List.contains s unreserveWords then
-            // You can assume this condition never fires - this is a deprecated, undocumented command line option anyway, we can ignore it.
-            IdentifierToken args lexbuf s
-        else
-          let mutable v = Unchecked.defaultof<_>
-          if keywordTable.TryGetValue(s, &v) then 
-              if (match v with RESERVED -> true | _ -> false) then
-                  warning(ReservedKeyword(FSComp.SR.lexhlpIdentifierReserved(s), lexbuf.LexemeRange));
-                  IdentifierToken args lexbuf s
-              else v
-          else 
+        match keywordTable.TryGetValue s with
```

Comment:
Please double check this return-a-tuple-and-match-on-it formulation produces code at least as good as before.  I still tend to write the mutable by hand in performance-critical code but I know there is an optimization that should give us the right code here.